sound/soc/sof/sof-priv.h

```diff
@@ -471,6 +471,9 @@ struct snd_sof_ipc {
 	/* disables further sending of ipc's */
 	bool disable_ipc_tx;
 
+	/* The current tx message is delayed due to congestion */
+	bool sending_delayed;
```

Comment:
it's not clear who delays and what this means. I find the comments lacking clarify on what this is supposed to fix.

Reply:
I think if you use the wording 'delay' there's an implicit notion of a workqueue or deferred processing.

In this case it's really 'send_on_next_irq'. there's no delay that's defined,

Reply:
My thinking was that we are delaying the sending of the message until the channel got free.

Initially I added the flag to `struct sof_intel_hda_dev` but it was a bit out of place, `struct sof_ipc4_fw_data` is completely out of scope for this kind of runtime handling.

I think I can move the flag down to `struct sof_intel_hda_dev` and name it differently? But still, the sending is delayed and I don't want to tie it to something concrete to be more reusable.
`ipc_tx_blocked` (in HDA scope)? The code otherwise would be the same, we will just not touch core code at all.
If the channel is BUSY when we first try to send, we set that the `ipc_tx` is `blocked` and we will re-try when the ACK arrives.

Reply:
The point is that you don't really wait until the channel is free. You will try to send the message when an interrupt is received on the RX path, possibly for something completely unrelated such as a DSP notification (could even be sets of notifications).

Reply:
@plbossart, no, we are waiting for an ACK on the _tx_ channel, so that the tx channel would be free. I deliberately not try to re-send a message when we receive a message (reply or notification).
In my reading the message is delayed for the duration the channel becomes free.

And no, we are not going to try to send a message if we receive a notification, the tx will be only open if we have received a reply (or we have a timeout). Notification will not change this, but we can receive as many notifications as it can fit within the timeout window between we sent a message and received the reply to it.

Reply:
not following, you are trying to send in an IRQ thread which could be triggered by the RX channel. We only have one IRQ...

Reply:
Yes, we are going to send a message from the irq thread if it was not possible to send when we have tried.
The first version used busy loop polling before sending a message to wait for it becoming free.

For the sending of the message it does not matter if we are in an irq thread or not.